contracts/BevorProtocol.sol

```diff
@@ -239,136 +241,165 @@ contract BevorProtocol is Ownable, ReentrancyGuard {
-        require(msg.sender == targetAudit.protocolOwner, "Cannot propose that the audit is invalid if you are not the protocol owner");
-
-        // Your DAO proposal creation logic might look like this: TODO:
-        // Replace the following lines with your actual DAO proposal
-        // creation code
-        address[] memory targets = new address[](1); 
-        uint256[] memory values = new uint256[](1);
-        bytes[] memory calldatas = new bytes[](1);
+    /**
+     * @notice Cancels a proposal. We'll constrain to the deployer address for now.
+     * Puts struct in a state where a new proposal cannot be made, and we mark it as non-frozen
+     * @param auditId the audit identifier
+     */
+    function cancelProposal(uint256 auditId) public onlyOwner {
+      Audit storage targetAudit = audits[auditId];
+
+      // this condition catches both empty proposalID, and frozen state for existing proposalID
+      require(
+        IDAOProxy(dao).isWithdrawFrozen(targetAudit.invalidatingProposalId),
+        "Proposal is not in state where it can be cancelled"
+      );
+      require(
+        !IDAOProxy(dao).isVestingInvalidated(targetAudit.invalidatingProposalId),
+        "Proposal was already invalidated"
+      );
+
+      IDAOProxy(dao).setProposalFrozen(targetAudit.invalidatingProposalId, false);
+    }
 
-        targets[0] = address(this); 
-        values[0] = 0;
-        calldatas[0] = bytes(calldata1);
+    /**
+     * @notice Carries out an invalidation. Funds are returned to Protocol Owner. NFT is burned.
+     * We'll constrain to the deployer address for now.
+     * @param auditId the audit identifier
+     */
+    function invalidate(uint256 auditId) public onlyOwner {
+      Audit storage targetAudit = audits[auditId];
+      uint256[] storage targetSchedules = auditToVesting[auditId];
+
+      // under the hood will revert if withdraw is frozen.
+      IDAOProxy(dao).setProposalInvalidated(targetAudit.invalidatingProposalId, true);
+
+      // payout whatever remains from the vested funds to the auditor
+      // as the difference between audit price and total withdrawn funds by auditors.
+      uint256 totalWithdrawn = 0;
+      uint256 totalAmount = 0;
+      for (uint256 i = 0; i < targetSchedules.length; i++) {
+        VestingSchedule storage schedule = vestingSchedules[targetSchedules[i]];
+        totalWithdrawn = totalWithdrawn + schedule.withdrawn;
+        totalAmount = totalAmount + schedule.amount;
+      }
+      
+      // AND then the DAO + voters... TBD.
+      targetAudit.token.transfer(targetAudit.protocolOwner, totalAmount - totalWithdrawn);
 
-        // Assuming 'dao' is your DAO contract
-        targetAudit.invalidatingProposalId = IDAOProxy(dao).propose(targets, values, calldatas, "Proposal to cancel vesting for audit");
-    }
+      // burn the NFT representation of an Audit
+      IAudit(nft).burn(targetAudit.nftTokenId);
```

Comment:
It's in here

Reply:
@blakehatch burn call is in here already

Reply:
Oh nice ok I approve

Reply:
breh

Reply:
mergin